### PR TITLE
Rewrites using packit

### PR DESCRIPTION
**Pull request by ryanmoran**
- uses https://github.com/ForestEckhardt/no-app-cnb to enable integration testing
- simplifies package structure by moving everything into the root
- introduces pattern to remove executable "bloat"

### CODE REVIEW

File: detect.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+				Name:    GoDependency,
+				Version: version,
+				Metadata: BuildPlanMetadata{
+					VersionSource: "buildpack.yml",
+					Build:         true,
+				},
+			})
+		}
+
+		return packit.DetectResult{
```
Comment (dwillist) [inline detect.go:38]: This is a departure from the way detect previously worked and breaks the `go-mod` integration tests.
This buildpack used to always require itself as a `build` dependency as seen [here](https://github.com/paketo-buildpacks/go-compiler/blob/8fa99f2a005a609d4599972598f4b285671cbb53/cmd/detect/main.go#L53).

The behavior implemented in this re-write seems preferable to the old behavior, but we should update the go-mod buildpack to require `go` as a build time dependency first [here](https://github.com/paketo-buildpacks/go-mod/blob/26260ef0b41674eca45c3ea61ffd4abe16109ed1/cmd/detect/main.go#L51)

File: main.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	"time"
+
+	"github.com/cloudfoundry/packit"
+	"github.com/cloudfoundry/packit/cargo"
+	"github.com/cloudfoundry/packit/postal"
+)
+
+func main() {
+	switch filepath.Base(os.Args[0]) {
+	case "detect":
```
Comment (dwillist) [inline main.go:16]: Cool beans 😎

File: detect.go (hunk 1 above)
Comment (ryanmoran) [reply detect.go:38]: I've added https://github.com/paketo-buildpacks/go-mod/pull/36 and https://github.com/paketo-buildpacks/dep/pull/25 to resolve this.

**Comment by ryanmoran** [comment]
> Why do you feel moving everything to the top level is a better design? For me it makes the flow harder to follow than when is was partitioned off. But that might just be me.

For me, its a few things:
1. we have quite a few repos that stutter (npm/npm, dep/dep, httpd/httpd, nginx/nginx, icu/icu)
2. the location of the program code changes in every repo, its jarring to have to remember what that sub-directory is called in each place
3. it makes it a bit clearer that these codebases are not really supposed to be imported as libraries (because they are main packages)

**Comment by ForestEckhardt** [comment]
Is there a structure that we could adopt that would allow us to separate the functional Go code from various other files that live in our directory such as the `README` or the `buildpack.toml`? It feels strange when I am looking at the buildpacks that these files are interspersed amongst bits of functional code. I agree with you stutter stepping point and non-uniform layout points and if this is just something I have to get used to then so be it, but potential alternative would be apperciated.

**Comment by ryanmoran** [comment]
Of course, we could have this in a named directory. I guess I just don't feel like its weird to have non-go files alongside go files. I'm not sure it matters what this named directory is called as long as it is uniform.

**Comment by ForestEckhardt** [comment]
@ryanmoran I am not strongly opinionated about this, I think I will give it time to sit and give this structure a shot.

**Pull request merged**